File: python/network_delay_time.py
import heapq
class Solution(object):
    # O(E * log(V)) time | O(V + E) space
    def networkDelayTime(self, times, n, k):
        """
        :type times: List[List[int]]
        :type n: int
        :type k: int
        :rtype: int
        """
        adj = {}
        for i in range(1, n + 1):
            adj[i] = []
        
        for time in times:
            s = time[0]
            d = time[1]
            w = time[2]
            adj[s].append([d, w])
        
        minheap = [(0, k)]
        costs = [float("infinity")] * (n + 1)
        costs[k] = 0
        visited = set()
        
        maxTime = 0
        print(adj)
        while minheap:
            w1, n1 = heapq.heappop(minheap)
            if n1 in visited:
                continue
            visited.add(n1)
            maxTime = max(maxTime, w1)

            neighbors = adj[n1]
            print(neighbors)
            for nei in neighbors:
                n2 = nei[0]
                w2 = nei[1]
                if w1 + w2 < costs[n2]:
                    costs[n2] = w1 + w2
                    heapq.heappush(minheap, (w1 + w2, n2))
        
        return maxTime if len(visited) == n else -1

times = [[2,1,1],[2,3,1],[3,4,1]]
n = 4
k = 2

solution = Solution()
print(solution.networkDelayTime(times, n, k))

File: src/main/java/august_2025/NetworkDelayTime.java
package august_2025;

import java.util.*;

public class NetworkDelayTime {

    public static void main(String[] args) {
//        int[][] times = {{2,1,1},{2,3,1},{3,4,1}};
//        int n = 4;
//        int k = 2;

        int[][] times = {{1,2,1},{2,3,2},{1,3,2}};
        int n = 3;
        int k = 1;

        NetworkDelayTime networkDelayTime = new NetworkDelayTime();
        int result = networkDelayTime.networkDelayTime(times, n, k);
        System.out.println(result);
    }

    // O(E * log(V)) time | O(V + E) space
    public int networkDelayTime(int[][] times, int n, int k) {
        Map<Integer, List<int[]>> adj = new HashMap<>();
        for (int i = 1; i <= n; i++) {
            adj.put(i, new ArrayList<>());
        }

        for (int[] time : times) {
            int s = time[0];
            int d = time[1];
            int w = time[2];
            adj.get(s).add(new int[] {d, w});
        }

        int maxTime = 0;
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        minHeap.add(new int[] {k, 0});
        Set<Integer> visited = new HashSet<>();

        int[] distances = new int[n + 1];
        Arrays.fill(distances, Integer.MAX_VALUE);
        distances[k] = 0;

        while (!minHeap.isEmpty()) {
            int[] current = minHeap.poll();
            int currNode = current[0];
            int w = current[1];
            if (visited.contains(currNode)) {
                continue;
            }
            visited.add(currNode);

            maxTime = Math.max(maxTime, w);
            List<int[]> neighbors = adj.get(currNode);
            for (int[] nei : neighbors) {
                int neiNode = nei[0];
                if (visited.contains(neiNode)) {
                    continue;
                }
                int neiWeight = nei[1];
                if (w + neiWeight < distances[neiNode]) {
                    distances[neiNode] = w + neiWeight;
                    minHeap.add(new int[] {neiNode, w + neiWeight});
                }
            }
        }
        return visited.size() == n ? maxTime : -1;
    }


}
